src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.SecKeyRef.cs

```diff
@@ -0,0 +1,275 @@
+                out signature,
+                out error);
+
+            using (error)
+            using (signature)
+            {
+                if (ret == 1)
+                {
+                    return CoreFoundation.CFGetData(signature);
+                }
+
+                if (ret == -2)
+                {
+                    throw CreateExceptionForCFError(error);
+                }
+
+                Debug.Fail($"GenerateSignature returned {ret}");
+                throw new CryptographicException();
+            }
+        }
+
+        internal static byte[] GenerateSignature(
+            SafeSecKeyRefHandle privateKey,
+            byte[] dataHash,
+            PAL_HashAlgorithm hashAlgorithm)
+        {
+            Debug.Assert(privateKey != null, "privateKey != null");
+            Debug.Assert(dataHash != null, "dataHash != null");
+            Debug.Assert(hashAlgorithm != PAL_HashAlgorithm.Unknown, "hashAlgorithm != PAL_HashAlgorithm.Unknown");
+
+            SafeCFDataHandle signature;
+            SafeCFErrorHandle error;
+
+            int ret = AppleCryptoNative_GenerateSignatureWithHashAlgorithm(
+                privateKey,
+                dataHash,
+                dataHash.Length,
+                hashAlgorithm,
+                out signature,
+                out error);
+
+            using (error)
+            using (signature)
+            {
+                if (ret == 1)
+                {
+                    return CoreFoundation.CFGetData(signature);
+                }
+
+                if (ret == -2)
+                {
+                    throw CreateExceptionForCFError(error);
+                }
+
+                Debug.Fail($"GenerateSignature returned {ret}");
+                throw new CryptographicException();
+            }
+        }
+
+        internal static bool VerifySignature(
+            SafeSecKeyRefHandle publicKey,
+            byte[] dataHash,
+            byte[] signature)
+        {
+            Debug.Assert(publicKey != null, "publicKey != null");
+            Debug.Assert(dataHash != null, "dataHash != null");
+            Debug.Assert(signature != null, "signature != null");
+
+            SafeCFErrorHandle error;
+
+            int ret = AppleCryptoNative_VerifySignature(
+                publicKey,
+                dataHash,
+                dataHash.Length,
+                signature,
+                signature.Length,
+                out error);
+
+            using (error)
+            {
+                if (ret == 1)
+                {
+                    return true;
+                }
+
+                if (ret == 0)
+                {
+                    return false;
+                }
+
+                if (ret == -2)
+                {
+                    throw CreateExceptionForCFError(error);
+                }
+
+                Debug.Fail($"VerifySignature returned {ret}");
+                throw new CryptographicException();
```

Comment:
This sequence appears to be fairly common.  Maybe you could refactor it out into a helper that's then used everywhere?